pints/plot/_surface.py

```diff
@@ -0,0 +1,324 @@
+#
+# Plots a function defined on a two-dimensional space.
+#
+# This file is part of PINTS (https://github.com/pints-team/pints/) which is
+# released under the BSD 3-clause license. See accompanying LICENSE.md for
+# copyright notice and full license details.
+#
+# This module was was adapted from Myokit (see http://myokit.org) by the
+# original author.
+#
+# The code to plot voronoi regions was based on an example shown here:
+# http://stackoverflow.com/a/20678647/423420
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+
+import numpy as np
+
+import pints
+
+
+def surface(points, values, boundaries=None, markers='+', figsize=None):
+    """
+    Takes irregularly spaced points and function evaluations in a
+    two-dimensional parameter space and creates a coloured surface plot using a
+    voronoi diagram.
+
+    Returns a ``matplotlib`` figure object and axes handle.
+
+    Parameters
+    ----------
+    points
+        A list of (two-dimensional) points in parameter space.
+    values
+        The values corresponding to these points.
+    boundaries
+        An optional :class:`pints.RectangularBoundaries` object to restrict the
+        area shown.
+    markers
+        The markers to use to plot the sampled points. Set to ``None`` to
+        disable.
+
+    """
+    import matplotlib.pyplot as plt
+    from matplotlib.collections import PolyCollection
+
+    # Check points and values
+    points = pints.matrix2d(points)
+    n, d = points.shape
+    if d != 2:
+        raise ValueError('Only two-dimensional parameters are supported.')
+    values = pints.vector(values)
+    if len(values) != n:
+        raise ValueError(
+            'The number of values must match the number of points.')
+
+    # Extract x and y points
+    x = points[:, 0]
+    y = points[:, 1]
+    del(points)
+
+    # Check boundaries
+    if boundaries is None:
+        xmin = 0.80 * np.min(x)
+        xmax = 1.02 * np.max(x)
+        ymin = 0.80 * np.min(y)
+        ymax = 1.02 * np.max(y)
+    else:
+        if boundaries.n_parameters() != 2:
+            raise ValueError(
+                'If given, the boundaries must be two-dimensional.')
+
+        xmin, ymin = boundaries.lower()
+        xmax, ymax = boundaries.upper()
+
+    # Get voronoi regions (and filter points and evaluations)
+    xlim = xmin, xmax
+    ylim = ymin, ymax
+    x, y, values, regions = _voronoi_regions(x, y, values, xlim, ylim)
+
+    # Create figure and axes
+    figure, axes = plt.subplots(figsize=figsize)
+    axes.set_xlim(xmin, xmax)
+    axes.set_ylim(ymin, ymax)
+
+    # Add coloured voronoi regions
+    c = PolyCollection(
+        regions, array=values, edgecolors='none', cmap='viridis_r')
+    axes.add_collection(c)
+
+    # Add markers
+    if markers:
+        axes.plot(x, y, markers)
```

Comment:
Specify the colour as well -- so that it matches (even if different version...?) the viridis_r cmap set above?